### PR TITLE
Add browser cookies support on Android

### PR DESCRIPTION
**Pull request by Fatxx**
**Description**

At the moment users are not able to use cookies on Android since the HTTP client (OkHttp) is [rejecting cookies by default](https://square.github.io/okhttp/4.x/okhttp/okhttp3/-cookie-jar/), this PR implements an interface that saves and loads cookies accordingly.

This PR also fixes clearing cookies on iOS and improves the "Clear Cookies" section structure.

**Test Cases**

Case 1 (Android):
 - Access a website that has cookies or login into one that has session cookies (ex.: Go to https://notareal.club/shop)
 - Add a product to your cart.
- Click view cart or navigate to the cart page.
- You should see an item in the cart

Cases 2 (iOS):
 - Access a website that has cookies or login into one that has session cookies (ex.: Go to https://notareal.club/shop)
 - Go to Settings -> Security & Privacy -> Clear Browser Cookies
 - Clear cookies button should be disabled
 - Navigate to the browser you should have no cookies

**Checklist**

* [x] There is a related GitHub issue
* [ ] Tests are included if applicable
* [ ] Any added code is fully documented

**Screenshots/Recordings**

https://user-images.githubusercontent.com/575479/178604012-9db220e4-f3a1-4aca-8f98-f2a10269f9a3.mp4

https://user-images.githubusercontent.com/575479/178604025-0a7f9f5d-cc4d-4a1f-bf00-5ba84a17dae2.mp4

https://user-images.githubusercontent.com/575479/178604034-9f017d65-fa62-450d-b80f-d7318dcf36e8.mp4

**Issue**

Progresses #4245, #2954, #2544, #4190

### CODE REVIEW

File: app/components/Views/Settings/SecuritySettings/Sections/ClearCookiesSection.tsx [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+const ClearCookiesSection = () => {
+  const [cookiesModalVisible, setCookiesModalVisible] = useState(false);
+  const [hasCookies, setHasCookies] = useState(false);
+  const { colors } = useAppThemeFromContext() || mockTheme;
+  const styles = createStyles(colors);
+
+  useEffect(() => {
+    const run = async () => {
+      if (Platform.OS === 'android') {
```
Comment (gantunesr) [inline app/components/Views/Settings/SecuritySettings/Sections/ClearCookiesSection.tsx:64]: Use the `Device` module from `utils`

File: app/components/Views/Settings/SecuritySettings/Sections/ClearCookiesSection.tsx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        // CookieManager.getAll() is not supported on Android
+        setHasCookies(true);
+      }
+
+      if (Platform.OS === 'ios') {
```
Comment (gantunesr) [inline app/components/Views/Settings/SecuritySettings/Sections/ClearCookiesSection.tsx:69]: Same here. Use the `Device` module from `utils`

File: app/components/Views/Settings/SecuritySettings/Sections/ClearCookiesSection.tsx [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+  const toggleClearCookiesModal = () =>
+    setCookiesModalVisible(!cookiesModalVisible);
+
+  const clearCookies = async () => {
+    const useWebKit = true;
+    await CookieManager.clearAll(useWebKit);
+    Logger.log('Browser cookies cleared');
+
+    if (Platform.OS === 'ios') {
```
Comment (gantunesr) [inline app/components/Views/Settings/SecuritySettings/Sections/ClearCookiesSection.tsx:87]: Same here. Use the `Device` module from `utils`

File: app/components/Views/Settings/SecuritySettings/Sections/ClearCookiesSection.tsx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+      const cookies = await CookieManager.getAll(useWebKit);
+      setHasCookies(Object.keys(cookies).length > 0);
+    }
+
+    toggleClearCookiesModal();
+  };
+
+  return (
+    <>
+      <View style={styles.setting} testID={'clear-cookies-section'}>
```
Comment (gantunesr) [inline app/components/Views/Settings/SecuritySettings/Sections/ClearCookiesSection.tsx:97]: Lets move this constant to the `test-ids` file

File: app/components/Views/Settings/SecuritySettings/index.js [hunk 5]
```diff
@@ -1200,13 +1142,12 @@ class Settings extends PureComponent {
...
           {this.renderClearPrivacySection()}
           {this.renderClearBrowserHistorySection()}
-          {this.renderClearCookiesSection()}
           {this.renderPrivacyModeSection()}
           {this.renderMetaMetricsSection()}
           {this.renderThirdPartySection()}
           {this.renderApprovalModal()}
           {this.renderHistoryModal()}
-          {this.renderCookiesModal()}
+          <ClearCookiesSection />
```
Comment (gantunesr) [inline app/components/Views/Settings/SecuritySettings/index.js:1150]: Nice

**Review by gantunesr — CHANGES_REQUESTED** [review]
Some nits in the js part of the PR. @sethkfman can provide a better review on the Java part

**Comment by gantunesr** [comment]
The `priority` and `severity` labels should be used in issues and not PRs

File: patches/react-native-webview+11.13.0.patch [hunk 6]
```diff
@@ -453,6 +457,83 @@ index f743bbc..e88f145 100644
...
      public void onHostResume() {
        // do nothing
+@@ -1737,3 +2042,75 @@ public class RNCWebViewManager extends SimpleViewManager<WebView> {
+     }
+   }
+ }
++
++class CustomCookieJar implements CookieJar {
++  private CookieSyncManager mCookieSyncManager;
++  private static final boolean USES_LEGACY_STORE = Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP;
```
Comment (tommasini) [inline patches/react-native-webview+11.13.0.patch:467]: Check if legacy support is needed

File: patches/react-native-webview+11.13.0.patch [hunk 7]
```diff
@@ -453,6 +457,83 @@ index f743bbc..e88f145 100644
...
++
++  public CustomCookieJar (WebView webView) {
```
Comment (tommasini) [inline patches/react-native-webview+11.13.0.patch:469]: Should move logic to the outside of web view file

File: app/components/Views/Settings/SecuritySettings/Sections/ClearCookiesSection.tsx [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+const ClearCookiesSection = () => {
+  const [cookiesModalVisible, setCookiesModalVisible] = useState(false);
+  const [hasCookies, setHasCookies] = useState(false);
+  const { colors } = useAppThemeFromContext() || mockTheme;
+  const styles = createStyles(colors);
+
+  useEffect(() => {
+    const run = async () => {
+      if (Device.isAndroid()) {
+        // CookieManager.getAll() is not supported on Android
```
Comment (tommasini) [inline app/components/Views/Settings/SecuritySettings/Sections/ClearCookiesSection.tsx:67]: Remove the comment

**Comment by sethkfman** [comment]
Reviewed during meeting. Once the comments are addressed the PR looks good.

**Comment by chrisleewilcox** [comment]
Please resolve conversations, changes and conflicts.  Once resolved please removed `needs-dev-review` label.

**Comment by Fatxx** [comment]
Add as draft in favour of merging https://github.com/MetaMask/metamask-mobile/pull/3905

**Review by gantunesr — APPROVED** [review]
LGTM

**Comment by cortisiko** [comment]
@Fatxx A couple of things:

1.
 I am using the URL mentioned in [this](https://github.com/MetaMask/metamask-mobile/issues/4245#issuecomment-1125049113) comment to test and I can still reproduce the issue. that is, I switch to polygon  > open https://matic.art/  > tap connect > connect my wallet > I am taken back to the dapp home screen and my wallet is still not connected.
see here: http://recordit.co/nxXZBUUsqc

2.
this link doesn't work: https://notareal.club/shop 
I am taken to a blank page. 

3.

I am unable to log into a WordPress site

**Review by cortisiko — CHANGES_REQUESTED** [review]
@Fatxx sending this back your way:


**Issue 1**
I am experiencing a crash on android when I try to sign in using Github on [[gitcoin.co](http://gitcoin.co/)](http://gitcoin.co/)
As soon as I tap on sign up the app crashes. I tried to view the logs but i didn’t find anything useful. Unfortunately, I can’t continuously reproduce because of Recaptcha. Can you try to reproduce on your end as well?
**The device used**: xaomi redmi note 8 running android 11
See recording [here](http://recordit.co/wQdTgyibMu )

**Issue 2**
I am having an issue logging into [[wordpress.org](http://wordpress.org/)](http://wordpress.org/) on android. Everything works as expected on ios

to see screen recording view our slack chat [here](https://consensys.slack.com/archives/C035KHUJLHG/p1659397318187869?thread_ts=1659397284.705659&cid=C035KHUJLHG)

**Issue 3**
On any android device, I can’t add items to my cart when I go to this woocomerce website:  https://porterandyork.com/product-category/buy-beef-online/
see [android](http://recordit.co/3qEotCy7oO)
it works on [ios](http://recordit.co/NckwugI15G)

**Comment by cortisiko** [comment]
@Fatxx The issues I reported [here](https://github.com/MetaMask/metamask-mobile/pull/4407#pullrequestreview-1061122450) were fixed…. BUT I noticed you merged 3905 into your PR. This prompted me to test the sites listed in this issue: https://github.com/MetaMask/metamask-mobile/issues/3857

My findings are both Android-specific:

**1/**
Using this WebSocket website: https://www.hawku.com/ 
I am unable to get to the connect wallet view. I am stuck on the Recaptcha page. It seems as if the page keeps on refreshing even though I completed the Recaptcha. 

see here: http://recordit.co/qdztA4bc1Z 

**2/**
Using this WebSocket website: https://fly.io/launch/livebook

 I am prompted with an oops something went wrong view after signing up making the browser feel a bit unstable. 

See here: http://recordit.co/Rm28GuxA4I

**Review by cortisiko — APPROVED** [review]
@Fatxx This guy is QA passed 🌮 🌮 🌮 

Outside of the scenarios listed in the ticket description I also tested the following:

- https://github.com/MetaMask/metamask-mobile/issues/1931

- https://github.com/MetaMask/metamask-mobile/issues/1956


- Ensure that you can interact with https://matic.art/

- With https://porterandyork.com/product-category/buy-beef-online/ or any other woo-commerce you can:
        Add items to your cart
        log in/create an account 
       And get to the check-out page

- You should be able to log into WordPress 
- You can go to http://gitcoin.co/ and proceed to log in/sign up with GitHub

**Pull request merged**